educe/rst_dt/learning/features.py

```diff
@@ -2,390 +2,243 @@
+        yield ('word_first', raw_words[0])
+        yield ('word_last', raw_words[-1])
+    if len(raw_words) > 1:
+        yield ('bigram_first', (raw_words[0], raw_words[1]))
+        yield ('bigram_last', (raw_words[-2], raw_words[-1]))
+    yield ('num_tokens', len(raw_words))
 
 
 # PTB tokens
-
-@ptb_tokens_feature
-@on_first_unigram
-def ptb_pos_tag_first(token):
-    "POS tag for first PTB token in the EDU"
-    # note: PTB tokens may not necessarily correspond to words
-    return token.tag
-
-
-@ptb_tokens_feature
-@on_last_unigram
-def ptb_pos_tag_last(token):
-    "POS tag for last PTB token in the EDU"
-    # note: PTB tokens may not necessarily correspond to words
-    return token.tag
-
-
-@ptb_tokens_feature
-@on_first_unigram
-def ptb_word_first(token):
-    "first PTB word in the EDU"
-    return token.word
-
-
-@ptb_tokens_feature
-@on_last_unigram
-def ptb_word_last(token):
-    "last PTB word in the EDU"
-    return token.word
-
-
-@ptb_tokens_feature
-@on_first_bigram
-def ptb_pos_tag_first2(token):
-    "POS tag for first two PTB tokens in the EDU"
-    # note: PTB tokens may not necessarily correspond to words
-    return token.tag
-
-
-@ptb_tokens_feature
-@on_last_bigram
-def ptb_pos_tag_last2(token):
-    "POS tag for last two PTB tokens in the EDU"
-    # note: PTB tokens may not necessarily correspond to words
-    return token.tag
-
-
-@ptb_tokens_feature
-@on_first_bigram
-def ptb_word_first2(token):
-    "first two PTB words in the EDU"
-    return token.word
-
-
-@ptb_tokens_feature
-@on_last_bigram
-def ptb_word_last2(token):
-    "last PTB words in the EDU"
-    return token.word
+# note: PTB tokens may not necessarily correspond to words
+
+SINGLE_PTB_TOKEN_WORD = [
+    ('ptb_word_first', Substance.DISCRETE),
+    ('ptb_word_last', Substance.DISCRETE),
+    ('ptb_word_first2', Substance.DISCRETE),
+    ('ptb_word_last2', Substance.DISCRETE)
+]
+
+
+def extract_single_ptb_token_word(edu_info):
+    """word features on PTB tokens for the EDU"""
+    try:
+        words = edu_info['words']
+    except KeyError:
+        return
+
+    if words:
+        yield ('ptb_word_first', words[0])
+        yield ('ptb_word_last', words[-1])
+
+    if len(words) > 1:
+        yield ('ptb_word_first2', (words[0], words[1]))
+        yield ('ptb_word_last2', (words[-2], words[-1]))
+
+
+SINGLE_PTB_TOKEN_POS = [
+    ('ptb_pos_tag_first', Substance.DISCRETE),
+    ('ptb_pos_tag_last', Substance.DISCRETE),
+    ('ptb_pos_tag_first2', Substance.DISCRETE),
+    ('ptb_pos_tag_last2', Substance.DISCRETE),
+]
+
+
+def extract_single_ptb_token_pos(edu_info):
+    """POS features on PTB tokens for the EDU"""
+    try:
+        tags = edu_info['tags']
+    except KeyError:
+        return
+
+    if tags:
+        yield ('ptb_pos_tag_first', tags[0])
+        yield ('ptb_pos_tag_last', tags[-1])
+
+    if len(tags) > 1:
+        yield ('ptb_pos_tag_first2', (tags[0], tags[1]))
+        yield ('ptb_pos_tag_last2', (tags[-2], tags[-1]))
+
```

Comment:
Would it be worth a refactoring this with some sort of abstraction that returns either the first/last or first/last 2 of something?